src/proposal.mligo

```diff
@@ -134,10 +147,24 @@ let stake_tk(token_amount, addr, store : nat * address * storage): storage =
   in { store with freeze_history = new_freeze_history }
 
 [@inline]
-let unfreeze_on_ledger (tokens, addr, ledger, total_supply, unfrozen_token_id, frozen_token_id : nat * address * ledger * total_supply * token_id * token_id): (ledger * total_supply) =
+let unfreeze_on_ledger (tokens, addr, ledger, total_supply, frozen_token_id, governance_token : nat * address * ledger * total_supply * token_id * governance_token): (operation * ledger * total_supply) =
+  // Call transfer on token_contract to transfer `token` number of
+  // tokens from `addr` to the address of this contract.
+  let
+    transfer_param : transfer_params =
+      [{ from_ = Tezos.self_address; txs = [ { to_ = addr; token_id = governance_token.token_id; amount = tokens } ] }] in
+  let token_contract =
+    begin
+      match (Tezos.get_entrypoint_opt "%transfer" governance_token.address : ((transfer_params contract) option)) with
+        | Some (c) -> c
+        | None -> (failwith "BAD_TOKEN_CONTRACT" : (transfer_params contract))
+    end in
+  let operation = Tezos.transaction transfer_param 0mutez token_contract in
```

Comment:
A good part of this code is duplicated here in `freeze_on_ledger` as well as in `transfer_contract_tokens` (in `token.mligo`).

We should abstract this to a function, it being used in several places will cause Ligo to compile it into a lambda and save some costs.

IMO we can replace the `FAIL_TRANSFER_CONTRACT_TOKENS` error in `transfer_contract_tokens` with `BAD_TOKEN_CONTRACT` as well as they effectively are the same thing (remember to update the spec tho).

Reply:
Have changed.